Gutenboarding: Extract each step into a separate component

razvanpapadopol opened the pull request:
#### Changes proposed in this Pull Request
- add `activeStep` to local state
- use consistent handlers to track step progress
- extract steps to separate files
- split CSS to live closer to the related components
- cleanup `edit` file by extracting question wrapper to the `Question` component

#### Testing instructions
- interacting with the Vertical Selector should feel more responsive
- when in collapsed mode (smaller font size), Vertical Selector display the previous selected value, instead of the last query

#### To do:
- [x] extract each step component to its own file 
- [x] extract duplicated HTML markup to a `Question` component

Commit:
Extract each step into a React component.
- add activeStep to local state
- use consistent handlers to track step progress

Commit:
Extract SiteType select to a separate file.
- also extract siteTypeOptions to site-type-select file

Commit:
Extract SiteTitle to a separate component.
- Also make the empty state UX consistent.

ockham commented on client/landing/gutenboarding/onboarding-block/edit.tsx:
```diff
@@ -3,155 +3,134 @@
 
-export default function OnboardingEdit() {
-	const [ lastSelectedQuestion, setLastSelectedQuestion ] = useState( null );
+const Step1 = ( { isActive, onExpand, onSelect }: StepProps ) => {
```
Nitpick, let's give them descriptive names -- the order might be subject to change, but the functionality will likely remain the same

```suggestion
const SiteTypeStep = ( { isActive, onExpand, onSelect }: StepProps ) => {
```

razvanpapadopol replied:
Thanks for the suggestion!
They have their own names and files now.

Commit:
Extract generic step related HTML to a new Question component.
- separate CSS to relevant components
- share the StepInputProps interface between separate step components

Commit:
- Fix typescript and linting errors.
- Remove useCallback where not necessary

Commit:
Fix store path.

ockham commented on client/landing/gutenboarding/onboarding-block/edit.tsx:
```diff
@@ -3,156 +3,74 @@
+			label: isFilledFormValue( siteType )
+				? NO__( 'I want to create a website ' )
+				: NO__( 'Setting up a website ' ),
+			displayValue: isFilledFormValue( siteType ) ? siteTypeOptions[ siteType ] : '',
```
This information is tightly coupled to the `SiteTypeSelect` component and shouldn't live outside of it -- we're basically splitting the source of truth for infomation pertaining to this step, which is a rather confusing pattern.

If we want to use a generic `Question` component to be re-used for each question, the pattern should be that each of them uses the `Question` component via composition, not the other way round.

ockham replied:
I _think_ that some of this might be coming from how we're iterating over `steps` here, and render a `Question` component for each. Let's ditch that, and use `React.Children.map()` instead, and just wrap the steps in `OnboardingEdit` (or possibly an intermediary component). Something like:

```js
<StepperWizard>
    <SiteTypeSelect />
    <VerticalSelect />
    <SiteTitle />
```

ockham commented on client/landing/gutenboarding/onboarding-block/edit.tsx:
```diff
@@ -3,156 +3,74 @@
+					( step, index ) =>
+						( index < 1 || isFilledFormValue( steps[ index - 1 ].value ) ) && (
+							<Question
+								key={ index }
```
https://medium.com/@robinpokorny/index-as-a-key-is-an-anti-pattern-e0349aece318 :grimacing:

razvanpapadopol replied:
In this case, our steps are statically defined and even if it may seem we are filtering them, they are always displayed in the same order so I'm pretty sure React will not get crazy.

This kind of exception is also mentioned in the article above:
> - the list and items are static–they are not computed and do not change;
> - the items in the list have no ids;
> - the list is never reordered or filtered.
>
> When all of them are met, you may safely use the index as a key.

ockham commented on client/landing/gutenboarding/onboarding-block/site-type-select/style.scss:
```diff
@@ -0,0 +1,67 @@
+		color: $light-gray-800;
+	}
+	
+	li.selected	label:hover::after,
```
Hmm we should fix that whitespace :confused:

Commit:
Replace tab with space.

Commit:
Add StepperWizard component to inject controls into step components.

ockham commented:
Styling seems to have changed a bit:

![image](https://user-images.githubusercontent.com/96308/68860879-93490d00-0714-11ea-8a6f-b249658a4e3b.png)

compared to `master`:

![image](https://user-images.githubusercontent.com/96308/68860922-a78d0a00-0714-11ea-9905-3e4e58a55b31.png)

ockham commented on client/landing/gutenboarding/onboarding-block/edit.tsx:
```diff
@@ -2,163 +2,35 @@
+					<SiteTypeSelect />
+					{ isFilledFormValue( siteType ) && <VerticalSelect /> }
+					{ isFilledFormValue( siteVertical ) && <SiteTitle /> }
+				</StepperWizard>
```
:heart:

ockham commented on client/landing/gutenboarding/onboarding-block/question/index.tsx:
```diff
@@ -0,0 +1,45 @@
+	</div>
+);
+
+export default Question;
```
I'm not entirely sure if the `Question` component gives us that much value, or if we should just merge the logic into each individual step component. After all, it's basically just a ternary, isn't it? :thinking:

razvanpapadopol replied:
This component is handling the structure of each question (wrapper, static label, link to expand the question) and it's also in charge with generic styling. By extracting this stuff here we not only save some code duplication but also guarantee consistency and flexibility. I think this is what composition is about, right?

ockham commented on client/landing/gutenboarding/onboarding-block/site-title.tsx:
```diff
@@ -0,0 +1,44 @@
+import { StepInputProps } from './stepper-wizard';
+import Question from './question';
+
+export default function SiteTitle( { onSelect, inputClass, ...props }: StepInputProps ) {
```
Let's destructure props explicitly :grimacing: We've had some instances where we did `...props` in Calypso in cases like this (where we actually knew which props to expect, and basically just used it as a shorthand), and it didn't turn out to be good practice :grimacing: 

(Same for the other step components.)

ockham commented on client/landing/gutenboarding/onboarding-block/site-title.tsx:
```diff
@@ -0,0 +1,44 @@
+					placeholder={ NO__( 'enter a title' ) }
+					onChange={ handleChange }
+					onBlur={ onSelect }
+					value={ siteTitle }
```
The duplication of values that we pass both to `<Question />` _and_ to `<input />` here (the `placeholder` string, and `value`) kind of seems like evidence for [this hypothesis](https://github.com/Automattic/wp-calypso/pull/37491/files#r346310923) to me :thinking:

razvanpapadopol replied:
The placeholder of the input and the empty `displayValue` passed to `Question` are not necessary the same (otherwise we'd be just using a constant for both). The former should help the user better understand what to time inside the input (with examples maybe) while the later is more like a call to action.

ockham commented on client/landing/gutenboarding/onboarding-block/stepper-wizard.tsx:
```diff
@@ -0,0 +1,33 @@
+				inputClass: 'onboarding-block__question-input',
+			} )
+	);
+}
```
Love this!

ockham commented on client/landing/gutenboarding/onboarding-block/stepper-wizard.tsx:
```diff
@@ -0,0 +1,33 @@
+				isActive: index === activeStep,
+				onExpand: () => setActiveStep( index ),
+				onSelect: handleNext,
+				inputClass: 'onboarding-block__question-input',
```
Could this be the reason for [this style change](https://github.com/Automattic/wp-calypso/pull/37491#issuecomment-553886928)?

razvanpapadopol replied:
No, this was just a way to have the input class as a constant. I was tempted to introduce an `<QuestionInput />` component but I dropped the idea in the end since it wasn't doing more than adding this class.

ockham commented on client/landing/gutenboarding/onboarding-block/stepper-wizard.tsx:
```diff
@@ -0,0 +1,33 @@
+		props.children,
+		( child, index ) =>
+			child &&
+			React.cloneElement( child, {
```
Getting a type error here. I found a relevant [SO thread](https://stackoverflow.com/questions/42261783/how-to-assign-the-correct-typing-to-react-cloneelement-when-giving-properties-to).

I was trying

```suggestion
			React.cloneElement( child as React.ReactElement< StepInputProps >, {
```

but was still getting a TS error (that I don't really understand :sweat_smile: ).

Maybe @sirreal can help? :grimacing:

ockham reviewed (approved):
Thanks, this is looking pretty good!

I left a few notes. I think the changed styling is the only actual issue; once that's fixed, feel free to address or ignore my other suggestions, and :ship:

Commit:
Destructure props explicitly for each step.

sirreal commented:
Rebased.

sirreal committed:
Clean and simplify Question component

sirreal committed:
Type as FunctionComponent

sirreal committed:
Rework children map/clone for correctness

- Wrap in Fragment
- Use `React.isValidElement` type guard

razvanpapadopol commented on client/landing/gutenboarding/onboarding-block/stepper-wizard.tsx:
```diff
@@ -15,16 +15,20 @@ const StepperWizard: FunctionComponent = ( { children } ) => {
+			{ React.Children.map(
+				children,
+				( child, index ) =>
+					React.isValidElement( child ) &&
```
👍

sirreal committed:
Rework prop injection and use FunctionComponent to type

sirreal committed:
Use @wordpress/keycodes and add dep

sirreal committed:
Use createRef for component ref

sirreal committed:
Try to improve the ref

sirreal committed:
Use ellipsis

sirreal committed:
Use lock's keycodes version

sirreal committed:
Remove redundant type assertion

sirreal reviewed (approved):
Rebased and I've been tweaking this, but this is good work and ready to land.

Thanks @razvanpapadopol!

sirreal committed:
Fix site type color regression

sirreal commented:
I fixed the regression mentioned here https://github.com/Automattic/wp-calypso/pull/37491#issuecomment-553886928 with 758b87d4205a65c8342dd1e5a79ab8df8bbd001d.

Should be ready to land.
